### PR TITLE
Filter VPC by VNI

### LINKED ISSUE #1756: Update firewall rule host filters with type VPC
Follow-up to #1636: whenever an instance or NIC is added or deleted in some VPC, all firewall rules that have that VPC as a host filter must be updated. This is particularly important since the default firewall rule `allow-internal-inbound` uses just such a host filter.

### LINKED ISSUE #1821: Default firewall rules don't match up with RFD 21
Taking the latest omicron + OPTE for a spin, I noticed the firewall rules don't seem to line up with the defaults described in RFD 21 §2.8.1.

Here's what I see via opteadm (note this is using my latest dev branch, the "default deny" feature does not exist in the version of OPTE that Omicron is current using. In that version the default is to allow.

```
rpz@kalm:~/oxidecomputer/opte$ pfexec opteadm dump-layer -p opte0 firewall
Layer firewall
======================================================================
Inbound Flows
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     ACTION                

Outbound Flows
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     ACTION                

Inbound Rules [Default: deny]
----------------------------------------------------------------------
ID       PRI    PREDICATES                                       ACTION            
3        65534  inner.ip.proto=TCP                               "Stateful: Firewall"
                inner.ulp.dst=22                                

2        65534  inner.ip.proto=TCP                               "Stateful: Firewall"
                inner.ulp.dst=3389                              

1        65534  inner.ip.src=172.30.0.5                          "Stateful: Firewall"
0        65534  inner.ip.proto=ICMP                              "Stateful: Firewall"

Outbound Rules [Default: deny]
----------------------------------------------------------------------
ID       PRI    PREDICATES                                       ACTION
```

Below is a list of what we should have.

|Default Rule |Current Omicron |Omicron w/ latest OPTE |Notes |
|-------------|-----------------|------------------------|-------|
|Implied allow outbound | CORRECT | INCORRECT |Needs explicit allow now that OPTE has default deny. Actually, thinking about it more, perhaps the default outbound action for the firewall layer should be allow.|
|Implied deny inbound |INCORRECT |CORRECT |Current is missing an explicit rule to deny inbound. We get it for free with latest OPTE because it sets firewall to deny by default in both directions.|
|Default allow internal inbound |INCORRECT |INCORRECT |I think this is supposed to be rule id 1, but it assigned the guest's IP, instead of the subnet CIDR.|
|Default allow ssh |CORRECT |CORRECT |Covered by rule id 3.|
|Default allow ICMP |CORRECT |CORRECT |Covered by rule id 0.|
|Default allow RDP |CORRECT |CORRECT |Covered by rule id 2.|

So in general, anyone currently using Omicron should be fine, as their guest connectivity will remain how it's always been before the firewall rule plumbing landed. But as part of upgrading to the next version of OPTE we'll also want to update the firewall rules to get it 100% in line with RFD 21.

### PR DESCRIPTION
**Pull request by plotnick**
Fixes #1756 (using option 3) and fixes #1821 (using new OPTE default actions and proper VPC/VNI host filters). Depends on #1904.

### CODE REVIEW

**Comment by plotnick** [comment]
The test failure on `helios / deploy` (`BadApiVersion { user: 18, kernel: 15 }`) should go away as soon as #1904 is merged.

**Comment by rzezeski** [comment]
I'm not sure the `VpcAddress` name makes sense. As the hosts/address filter could be for a VPC Subnet address or any address at all. Also, based on RFD 21 table 14 it seems like the VPC "target/filter" is meant to always target the entire VPC. So when used as a firewall host filter it should probably always result in an `Address::Vni` value.

Do we need to update the `DEFAULT_FIREWALL_RULES` value to use the new VNI filter, or does it just do the right thing now based on what "default" means?

**Comment by plotnick** [comment]
> I'm not sure the `VpcAddress` name makes sense.

Yeah, I'm not wild about the name either. But we need an enum like this (though maybe called something different) that is either an IP address/network or a VNI indicating a VPC. Alternatively, we could (as OPTE did) push the VNI address type down into `IpNet`, but then _that_ would be poorly named, because it could then designate networks by non-IP-address means. So I prefer the former route, but am open to naming suggestions. `VpcOrIpNet`? `VpcNet`? `Net`?

> Do we need to update the `DEFAULT_FIREWALL_RULES` value to use the new VNI filter, or does it just do the right thing now based on what "default" means?

The latter (i.e., no update is needed). Any host filter of `"type": "vpc"` (e.g., the one for `default`) will be translated into a VNI. (As a target it does what it did before, which is to either apply or not apply the rule depending on whether the instance is in the VPC.) For instance, here's the result of `opteadm dump-layer firewall -p opte0` after booting a new instance with the default firewall rules:
```
Inbound Rules [Default: deny]
----------------------------------------------------------------------
ID       PRI    PREDICATES                                       ACTION            
2        65534  inner.ip.proto=TCP                               "Stateful Allow"
                inner.ulp.dst=22                                

1        65534  meta: vni=12897409                               "Stateful Allow"
0        65534  inner.ip.proto=ICMP                              "Stateful Allow"

Outbound Rules [Default: stateful allow]
----------------------------------------------------------------------
ID       PRI    PREDICATES                                       ACTION            
```
Rule ID 0 is `allow-icmp` (from anywhere), 1 is `allow-internal-inbound` (just from the VPC, identified by VNI), and 2 is `allow-ssh` (from anywhere).

**Comment by rzezeski** [comment]
> > I'm not sure the `VpcAddress` name makes sense.
> 
> Yeah, I'm not wild about the name either. But we need an enum like this (though maybe called something different) that is either an IP address/network or a VNI indicating a VPC. Alternatively, we could (as OPTE did) push the VNI address type down into `IpNet`, but then _that_ would be poorly named, because it could then designate networks by non-IP-address means. So I prefer the former route, but am open to naming suggestions. `VpcOrIpNet`? `VpcNet`? `Net`?
> 

Okay. For me I read `VpcAddress` as "this is a VPC address", but now I see you mean "this is a VPC or an address" (and I just noticed you stated this in the code comment). I think that was part of my confusion. I defer to others who spend more time working on the control plane. In fact, I think @bnaecker or someone else should also review this since I'm less familiar with some of the other bits here.

That said, this seems good to me.

**Review by rzezeski — APPROVED** [review]

File: common/src/api/external/mod.rs
Comment (bnaecker) [inline common/src/api/external/mod.rs:1285]: I think I understand why this enum exists, to cover the ways we ascribe a firewall rule to a set of hosts. Would a name more like `HostIdentifier` be appropriate?

I also wonder whether this needs to live in the external API. There are no external calls that should see this, AFAICT, so `common/src/api/internal/mod.rs` might be more appropriate. There may be some Rust visibility or Cargo dependency that makes that less attractive, but it might be nice to investigate.

**Review by bnaecker — APPROVED** [review]
Looks good, thanks! A suggestion to help resolve the naming issue @rzezeski brought up, but otherwise great.

File: common/src/api/external/mod.rs
Comment (plotnick) [reply common/src/api/external/mod.rs]: `HostIdentifier` is a great suggestion, thanks! And good call on the internal API; I always forget that we have that. Renamed and moved in 47361a5a.

(Minor follow-up question: after much playing around with building, rebuilding, and running the various `_openapi` tests with `EXPECTORATE=overwrite`, I still can't figure out why `HostIdentifier` shows up in `sled-agent.json` but not `nexus-internal.json`. The build seems happy, so I'm not going to sweat it, but I'm still a bit confused.)

Comment (bnaecker) [reply common/src/api/external/mod.rs]: That might be a question for @ahl! It's possible that types only show up in the OpenAPI output when they are actually needed to make requests to a given server. Since this type is technically only used by Nexus making calls _to_ the sled agent, only including needed names could explain it. But that's a WAG :)

**Pull request merged**
